Applications/spire/source/charting/cached_chart_model.cpp

```diff
@@ -0,0 +1,125 @@
+        }
+        return std::vector<Candlestick>(first_iterator, last_iterator);
+      });
+    }
+  }
+  return load_data({first, last});
+}
+
+connection CachedChartModel::connect_candlestick_slot(
+    const CandlestickSignal::slot_type& slot) const {
+  return m_chart_model->connect_candlestick_slot(slot);
+}
+
+QtPromise<std::vector<Candlestick>> CachedChartModel::load_data(
+    const ChartRange& data) {
+  return m_chart_model->load(data.m_start, data.m_end).then(
+    [=] (auto result) {
+      auto first = std::lower_bound(m_loaded_data.begin(),
+        m_loaded_data.end(), data.m_start, [] (const auto& index,
+            const auto& value) {
+          return index.GetStart() < value;
+        });
+      auto last = std::lower_bound(m_loaded_data.begin(),
+        m_loaded_data.end(), data.m_end, [] (const auto& index,
+            const auto& value) {
+          return index.GetStart() < value;
+        });
+      auto pos = m_loaded_data.end();
+      if(last != m_loaded_data.end()) {
+        pos = m_loaded_data.erase(first, last + 1);
+      } else {
+        pos = m_loaded_data.erase(first, last);
+      }
+      m_loaded_data.insert(pos, result.Get().begin(), result.Get().end());
+      auto [range_first, range_last] = range_search(data);
+      auto new_range = data;
+      if(range_first != m_ranges.end()) {
+        new_range.m_start = std::min((*range_first).m_start, new_range.m_start);
+      }
+      if(range_last != m_ranges.end()) {
+        new_range.m_end = std::max((*range_last).m_end, new_range.m_end);
+      }
+      auto range_pos = m_ranges.erase(range_first, range_last);
+      m_ranges.insert(range_pos, new_range);
+      return result.Get();
+  });
+}
+
+std::tuple<std::vector<CachedChartModel::ChartRange>::iterator,
+    std::vector<CachedChartModel::ChartRange>::iterator>
+    CachedChartModel::range_search(const ChartRange& range) {
+  if(m_ranges.empty()) {
+    return {m_ranges.end(), m_ranges.end()};
+  }
+  auto first = std::optional<std::vector<ChartRange>::iterator>();
+  auto last = std::optional<std::vector<ChartRange>::iterator>();
+  if(range.m_start < m_ranges.front().m_start) {
+    first = m_ranges.begin();
+    if(range.m_end < m_ranges.front().m_start) {
+      return {first.value(), first.value()};
+    }
+  }
+  if(range.m_end > m_ranges.back().m_end) {
+    last = m_ranges.end();
+    if(range.m_start > m_ranges.back().m_end) {
+      return {last.value(), last.value()};
+    }
+  }
+  for(auto loaded_range = m_ranges.begin(); loaded_range != m_ranges.end();
+      ++loaded_range) {
+    if((*loaded_range).m_start <= range.m_start &&
+        (*loaded_range).m_end >= range.m_start) {
+      first = loaded_range;
+    }
+    if((*loaded_range).m_start <= range.m_end &&
+        (*loaded_range).m_end >= range.m_end) {
+      last = loaded_range + 1;
+    }
+    if(!first.has_value() && range.m_start > (*loaded_range).m_end &&
+        range.m_start < (*(loaded_range + 1)).m_start) {
```

Comment:
How do you know ```loaded_range + 1``` is valid?

Reply:
On line 110, this only executes when P.end is within a range, so at the very most range +1 == R.end().

For the dereference on line 113: execution returns on line 98 if range.start > m_ranges.back().end, and line 106 assigns first if loaded_range is between m_ranges.back().start and m_ranges.back().end, so by line 113 loaded_range's index is no greater than m_ranges.back() - 1.

It's the same idea for line 117.